src/cocoa/toga_cocoa/widgets/webview.py

```diff
@@ -66,3 +68,78 @@ def evaluate(self, javascript):
         :type  javascript: ``str``
         """
         return self.native.stringByEvaluatingJavaScriptFromString(javascript)
+
+
+if "WKWebView" in vars():
+    class TogaWKWebView(WKWebView):
+        @objc_method
+        def webView_didFinish_(self, navi):
+            if self.interface.on_webview_load:
+            self.interface.on_webview_load(self.interface)
+
+        @objc_method
+        def acceptsFirstResponder(self) -> bool:
+            return True
+
+        @objc_method
+        def keyDown_(self, event) -> None:
+            print('in keyDown', event.keyCode)
+            if self.interface.on_key_down:
+                self.interface.on_key_down(event.keyCode, event.modifierFlags)
+
+        @objc_method
+        def touchBar(self):
+            # Disable the touchbar.
+            return None
+
+    class WebKit2WebView(Widget):
+        def create(self):
+            self.native = TogaWKWebView.alloc().init()
+            self.native.interface = self.interface
+
+            self.native.uiDelegate = self.native
+            self.native.navigationDelegate = self.native
+
+            self.add_constraints()
+
+        def get_dom(self):
+            return self.evaluate(
+                'document.getElementsByTagName("html")[0].outerHTML')
+
+        def set_url(self, value):
+            if value:
+                self.interface.url = value
+
+            request = NSURLRequest.requestWithURL(NSURL.URLWithString(
+                self.interface.url))
+            self.native.load(request)
+
+        def set_content(self, root_url, content):
+            self.native.loadHTMLString_baseURL_(
+                content, NSURL.URLWithString(root_url))
+
+        def set_user_agent(self, value):
+            self.native.customUserAgent = value
+
+        def evaluate(self, js_str):
+            fur = Future()  # type: concurrent.futures.Future
+
+            def when_finish(result, err):
+                if fur.done():
+                    return
+
+                if err:
+                    fur.set_exception(err)
+
+                else:
+                    fur.set_result(result)
+
+            self.native.evaluateJavaScript_completionHandler_(
+                js_str, when_finish)
+
+            return fur
```

Comment:
I'm unsure about the change in API to return a Future. I understand the need, but given that we have support for asyncio internally, why not use that, await the result, and keep the public API synchronous (and therefore easier to use)?

Reply:
The reason why I am using `concurrent.futures.Future` is that I am not quite sure the status of `asyncio` adoption in toga, and using `asyncio.Future` here will make it non thread safe. And I am pretty sure that there're still handlers that is not being scheduled with the event loop(by looking at the issue list).

As for `keeping the public API synchronous`, there is currently no way to approach it by using `asyncio`. There are only two ways of doing it: `Callback` and `Coroutine`.

Both of them are not synchronous and must be scheduled with the event loop. `Coroutine`(either `yield`/`yield from` or `async`/`await`) is a pseudo-synchronous grammar sugar to simplify the use of asynchronous, but it is still not synchronous and must be scheduled with the event loop.

Also it is not necessary to make `evaluate` a coroutine function. As long as it returns `Awaitable[T]`, you can `await` it and get `T` when it's ready. This can be satisfied by changing from `concurrent.futures.Future` to `asyncio.Future`. Or wrap `concurrent.futures.Future` with `asyncio.wrap_future`.

Reply:
`asyncio` is 100% integrated into Toga, so there's no problem using it.

As for the API; I asked the question [here](https://gist.github.com/freakboy3742/c6a67d6594f6e9c741ff4151a34ccc51), and posted it [on Twitter](https://twitter.com/freakboy3742/status/937205265551601664) [(originally here)](https://twitter.com/freakboy3742/status/936795876281876480); consensus seems to agree with you... which is frustrating.

@ncoghlan has said he has a possible approach for restructuring the API to avoid this problem; we'll have to wait and see what that is.

I'm not wild about returning Futures; if we're need to go async, I'd rather go all-in and make the method awaitable as-is, rather than expecting the user to wrap their head around Futures.

Reply:
Taking a more user-API centric view on things (vs my implementation centric comments on the gist), there are 3 reasonable behaviours you can offer here:

1. Synchronous on the main thread: risks becoming unresponsive if the JS evaluation is slow, but matches your existing API behaviour. The kind of future you want is `concurrent.futures.Future`, and then you can just block on it, waiting for the callback to fire. (If I understand correctly, this is what the current PR does).
2. Asynchronous on the main thread: return an awaitable as @futursolo suggests above (this could also be offered as a separate `evaluate_async` API)
3. Synchronous in a separate thread: this is the main option I cover in my gist. It has the nice property of allowing the handler code to treat it like a synchronous call without risking blocking the main thread. The downside is that all other widget operations *also* need to be delegated back to the main thread (or otherwise synchronised to prevent race conditions)

Reply:
@freakboy3742 Unfortunately, this is the truth when using `asyncio`. Asynchrony must be explicit. You idea is actually making the asynchrony implicit. The only library that I know deals well with that is `gevent`. However, it requires monkey patching, which I will definitely avoid. 

@ncoghlan 
1. No. `fur.result()` blocks the entire thread, including the event loop. It is just a simple dead lock if you are trying to use it. The only reason that I use it is because that I don't have to use `loop.call_soon_threadsafe` for `asyncio.Future`.
2. `asyncio.Future` is awaitable. You can just return an future object(This is not to be confused with `concurrent.futures.Future`.) from `WKWebView.evaluate()`. If you don't want the result, do not `await` it.
3. `asyncio` is not thread safe, and you have to create `WKWebView` on the UI thread. So...

`concurrent.futures.Future` works with threads, so `fur.result()` will block the thread to wait the result(sending from the other threads).

However, `asyncio.Future` works with coroutines, so it will not block the thread when calling `fur.result()` and it will raise an error to prevent the thread from being blocked. You should use `await` to block the current coroutine instead of the thread.

Using `asyncio.Future` here is definitely a better choice. I was thinking about thread safety at that point of time and now it seems a little bit silly.

Reply:
@futursolo If I understand how the current WebKit1 based API works (note: it's entirely possible I don't), then toga's WebView already risks blocking the UI thread while running JavaScript code in the HTML widget.

In that case, adding WebKit2 support needs to trade-off between:

1. Keeping any existing code working (as it's actually OK to retain control of the entire UI thread if the JS code being evaluated is non-blocking and generally quick to evaluate - those are the same constraints that already apply to any Python code run in a handler)
2. Enabling access to the non-blocking JS evaluation support in WebKit2
3. Minimising the likelihood of *accidentally* making the main UI thread unresponsive via an ill-considered JS evaluation request

Depending on the chosen API design priorities, there are a few possible options there:

* keep the current WebView as purely synchronous, add a new AsyncWebView
* make the default WebView asynchronous, add a new SyncWebView
* offer both sync and async methods on the WebView object itself (with a `_sync` or `_async` suffix, depending on which behaviour you choose as the default at the method naming level)